packages/flutter/lib/src/widgets/navigator.dart

```diff
@@ -1881,33 +1883,47 @@ class NavigatorState extends State<Navigator> with TickerProviderStateMixin {
   Future<T> pushAndRemoveUntil<T extends Object>(Route<T> newRoute, RoutePredicate predicate) {
     assert(!_debugLocked);
     assert(() { _debugLocked = true; return true; }());
-    final List<Route<dynamic>> removedRoutes = <Route<dynamic>>[];
-    while (_history.isNotEmpty && !predicate(_history.last)) {
-      final Route<dynamic> removedRoute = _history.removeLast();
-      assert(removedRoute != null && removedRoute._navigator == this);
-      assert(removedRoute.overlayEntries.isNotEmpty);
-      removedRoutes.add(removedRoute);
-    }
+    final Route<dynamic> precedingRoute = _history.last;
     assert(newRoute._navigator == null);
     assert(newRoute.overlayEntries.isEmpty);
-    final Route<dynamic> oldRoute = _history.isNotEmpty ? _history.last : null;
     newRoute._navigator = this;
     newRoute.install(_currentOverlayEntry);
     _history.add(newRoute);
+
+    Route<dynamic> oldRoute;
+    int removalIndex = 0;
+
+    for (Route<dynamic> route in _history) {
+      if (predicate(route)) {
+        oldRoute = route;
+        removalIndex = _history.indexOf(route) + 1;
+        break;
+      }
+    }
+
+    // Notify for newRoute
+    newRoute.didChangeNext(null);
+    for (NavigatorObserver observer in widget.observers)
+      observer.didPush(newRoute, precedingRoute);
+
     newRoute.didPush().whenCompleteOrCancel(() {
       if (mounted) {
-        for (Route<dynamic> route in removedRoutes)
-          route.dispose();
+
+        // Complete remove until predicate & notify
+        while (_history[removalIndex] != newRoute) {
+          final Route<dynamic> removedRoute = _history.removeAt(removalIndex);
+          assert(removedRoute != null && removedRoute._navigator == this);
+          assert(removedRoute.overlayEntries.isNotEmpty);
+          for (NavigatorObserver observer in widget.observers)
+            observer.didRemove(removedRoute, oldRoute);
```

Comment:
This seems to be calling didRemove on the bottommost route first and then going up to the topmost. I beleive this breaks the contract for didRemove:

> If multiple routes are being removed, [...] this method will be called once for each removed route, from the topmost route to the bottommost route.

Reply:
Fixed.